pkg/xds/translate.go

```diff
@@ -40,8 +41,186 @@ const (
+							Match:     TranslateRouteMatch(nil),
+							Decorator: &route.Decorator{Operation: DefaultOperation},
+							Action: &route.Route_Route{
+								Route: &route.RouteAction{
+									ClusterSpecifier: &route.RouteAction_Cluster{Cluster: cluster},
+								},
+							},
+						},
+					}},
+				})
+			}
+		}
+	}
+
+	return out
+}
+
+// MatchServiceHosts splits the virtual service hosts into services and literal hosts
+func MatchServiceHosts(in model.Config, serviceByName ServiceByName) ([]string, []*model.Service) {
+	rule := in.Spec.(*routingv2.RouteRule)
+	hosts := make([]string, 0)
+	services := make([]*model.Service, 0)
+	for _, host := range rule.Hosts {
+		if svc := serviceByName(host, in.ConfigMeta.Namespace); svc != nil {
+			services = append(services, svc)
+		} else {
+			hosts = append(hosts, host)
+		}
+	}
+	return hosts, services
+}
+
+// TranslateVirtualHost creates virtual hosts corresponding to a virtual service.
+func TranslateVirtualHost(in model.Config, serviceByName ServiceByName, subsetSelector SubsetSelector) []GuardedHost {
+	hosts, services := MatchServiceHosts(in, serviceByName)
+	serviceByPort := make(map[int][]*model.Service)
+	for _, svc := range services {
+		for _, port := range svc.Ports {
+			if port.Protocol.IsHTTP() {
+				serviceByPort[port.Port] = append(serviceByPort[port.Port], svc)
+			}
+		}
+	}
+
+	// if no services matched, then we have no port information -- default to 80 for now
+	// TODO: use match condition ports
+	if len(serviceByPort) == 0 {
+		serviceByPort[80] = nil
+	}
+
+	out := make([]GuardedHost, len(serviceByPort))
+	for port, services := range serviceByPort {
+		clusterNaming := TranslateDestination(serviceByName, subsetSelector, in.ConfigMeta.Namespace, port)
+		routes := TranslateRoutes(in, clusterNaming)
+		out = append(out, GuardedHost{
+			Port:     port,
+			Services: services,
+			Hosts:    hosts,
+			Routes:   routes,
+		})
+	}
+
+	return out
+}
+
+// TranslateDestination produces a cluster naming function using the config context.
+func TranslateDestination(
+	serviceByName ServiceByName,
+	subsetSelector SubsetSelector,
+	contextNamespace string,
+	defaultPort int) ClusterNaming {
+	return func(destination *routingv2.Destination) string {
+		// detect if it is a service
+		svc := serviceByName(destination.Name, contextNamespace)
+
+		// TODO: create clusters for non-service hostnames/IPs
+		if svc == nil {
+			return UnresolvedCluster
+		}
+
+		// default port uses port number
+		svcPort, _ := svc.Ports.GetByPort(defaultPort)
+		if destination.Port != nil {
+			switch selector := destination.Port.Port.(type) {
+			case *routingv2.PortSelector_Name:
+				svcPort, _ = svc.Ports.Get(selector.Name)
+			case *routingv2.PortSelector_Number:
+				svcPort, _ = svc.Ports.GetByPort(int(selector.Number))
+			}
+		}
+
+		if svcPort == nil {
+			return UnresolvedCluster
+		}
+
+		// use subsets if it is a service
+		labels := subsetSelector(svc, destination.Subset)
```

Comment:
Why do we need labels? Since you are redoing a whole bunch of stuff, please use the subset as the cluster name instead of using a concatenation of labels. It shows up nicely in dashboards. I dont think this is going to break anything new, as this whole code is already new.

Reply:
Cause v1 does it so, and we're not changing it at the same time. Again, I used pluggable functions precisely so that we can tweak it later.